### PR TITLE
Convert concatenation string to interpolation for System.Private.Xml

### PR DESCRIPTION
**Pull request by kronic**

### CODE REVIEW

File: src/libraries/System.Private.Xml/src/System/Xml/Dom/DocumentXmlWriter.cs [hunk 1]
```diff
@@ -441,7 +441,8 @@ private void CloseWithAppendAttribute()
                 if (offset != -1
                     && ((XmlAttribute)attrs.nodes[offset]).Specified)
                 {
-                    throw new XmlException(SR.Xml_DupAttributeName, attr.Prefix.Length == 0 ? attr.LocalName : string.Concat(attr.Prefix, ":", attr.LocalName));
+                    throw new XmlException(SR.Xml_DupAttributeName, attr.Prefix.Length == 0 ? attr.LocalName : $"{attr.Prefix}:{attr.LocalName}"
+                    );
```
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/Dom/DocumentXmlWriter.cs:445]:
```suggestion
                    throw new XmlException(SR.Xml_DupAttributeName, attr.Prefix.Length == 0 ? attr.LocalName : $"{attr.Prefix}:{attr.LocalName}");
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/ImportContext.cs [hunk 2]
```diff
@@ -156,7 +156,7 @@ internal StringCollection Warnings
                 else
                 {
                     Warnings.Add(SR.Format(SR.XmlMismatchSchemaObjects, item.GetType().Name, qname.Name, qname.Namespace));
-                    Warnings.Add("DEBUG:Cached item key:\r\n" + (string?)looks[cachedItem] + "\r\nnew item key:\r\n" + (string?)looks[item]);
+                    Warnings.Add($"DEBUG:Cached item key:\r\n{(string?)looks[cachedItem]}\r\nnew item key:\r\n{(string?)looks[item]}");
```
Comment (am11) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/ImportContext.cs:159]: Roslyn converts four arguments to string[] and then calls string.Concat with interpolation. While with `+` operator, it emits a single string.Concat call: [sharplab with nightly roslyn build](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKAwGYACbUgYUIG9DSHyyMcAGUgWRwAoX2APAJSkAvAD5SAIgQB7ANYBTGCCmkA1KX7rN2/gG56jEuTacsvU0NESAJNPlKVNfgF9Sztx8kGCLoA===).

String interpolation is not necessarily the best alternative in all cases.

Comment (stephentoub) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/ImportContext.cs:159]: > Roslyn converts four arguments to string[] and then calls string.Concat with interpolation. While with + operator, it emits a single string.Concat call: sharplab with nightly roslyn build.

That's a nightly build of the compiler but targeting .NET 5.  You'd see different results if you were targeting .NET 6.
https://devblogs.microsoft.com/dotnet/string-interpolation-in-c-10-and-net-6/

**Review by eiriktsarpalis — APPROVED** [review]
LGTM although might need another pair of eyes.

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlTextWriter.cs [hunk 3]
```diff
@@ -1592,8 +1592,7 @@ private void PopNamespaces(int indexFrom, int indexTo)
         private string GeneratePrefix()
         {
             int temp = _stack[_top].prefixCount++ + 1;
-            return "d" + _top.ToString("d", CultureInfo.InvariantCulture)
-                + "p" + temp.ToString("d", CultureInfo.InvariantCulture);
+            return $"d{_top.ToString("d", CultureInfo.InvariantCulture)}p{temp.ToString("d", CultureInfo.InvariantCulture)}";
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Core/XmlTextWriter.cs:1595]:
```suggestion
            return string.Create(CultureInfo.InvariantCulture, $"d{_top}p{temp}";
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/Compiler.cs [hunk 4]
```diff
@@ -92,7 +92,7 @@ internal TextWriter Source
 
         internal static string GetTempAssemblyName(AssemblyName parent, string? ns)
         {
-            return parent.Name + ".XmlSerializers" + (string.IsNullOrEmpty(ns) ? "" : $".{GetPersistentHashCode(ns)}");
+            return $"{parent.Name}.XmlSerializers{(string.IsNullOrEmpty(ns) ? "" : $".{GetPersistentHashCode(ns)}")}";
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/Compiler.cs:95]:
```suggestion
            return string.IsNullOrEmpty(ns) ?
                $"{parent.Name}.XmlSerializers" :
                $"{parent.Name}.XmlSerializers.{GetPersistentHashCode(ns)}";
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSchemaExporter.cs [hunk 5]
```diff
@@ -147,7 +147,7 @@ public string ExportAnyType(string? ns)
                     if (IsAnyType(schemaType, true, true))
                         return name;
                     i++;
-                    name = "any" + i.ToString(CultureInfo.InvariantCulture);
+                    name = $"any{i.ToString(CultureInfo.InvariantCulture)}";
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSchemaExporter.cs:150]:
```suggestion
                    name = string.Create(CultureInfo.InvariantCulture, $"any{i}");
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSchemas.cs [hunk 6]
```diff
@@ -614,8 +614,8 @@ private static string Dump(XmlSchemaObject o)
         private static string MergeFailedMessage(XmlSchemaObject src, XmlSchemaObject dest, string? ns)
         {
             string err = SR.Format(SR.XmlSerializableMergeItem, ns, GetSchemaItem(src, ns, null));
-            err += "\r\n" + Dump(src);
-            err += "\r\n" + Dump(dest);
+            err += $"\r\n{Dump(src)}";
+            err += $"\r\n{Dump(dest)}";
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSchemas.cs:618]:
```suggestion
            err += $"{Environment.NewLine}{Dump(src)}{Environment.NewLine}{Dump(dest)}";
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationReader.cs [hunk 7]
```diff
@@ -792,12 +792,11 @@ protected void ParseWsdlArrayType(XmlAttribute attr)
...
                 if (colon < 0)
                 {
-                    attr.Value = _r.LookupNamespace("") + ":" + attr.Value;
+                    attr.Value = $"{_r.LookupNamespace("")}:{attr.Value}";
                 }
                 else
                 {
-                    attr.Value = string.Concat(_r.LookupNamespace(attr.Value.Substring(0, colon)), ":",
-                        attr.Value.AsSpan(colon + 1));
+                    attr.Value = $"{_r.LookupNamespace(attr.Value.Substring(0, colon))}:{attr.Value.AsSpan(colon + 1)}";
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationReader.cs:799]:
```suggestion
                    attr.Value = $"{_r.LookupNamespace(attr.Value.AsSpan(0, colon))}:{attr.Value.AsSpan(colon + 1)}";
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationReader.cs [hunk 8]
```diff
@@ -2129,8 +2128,8 @@ internal Member(XmlSerializationReaderCodeGen outerClass, string source, string?
             internal Member(XmlSerializationReaderCodeGen outerClass, string source, string? arraySource, string? arrayName, int i, MemberMapping mapping, bool multiRef, string? choiceSource)
             {
                 _source = source;
-                _arrayName = arrayName + "_" + i.ToString(CultureInfo.InvariantCulture);
-                _choiceArrayName = "choice_" + _arrayName;
+                _arrayName = $"{arrayName}_{i.ToString(CultureInfo.InvariantCulture)}";
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationReader.cs:2131]:
```suggestion
                _arrayName = string.Create(CultureInfo.InvariantCulture, $"{arrayName}_{i}");
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriter.cs [hunk 9]
```diff
@@ -870,7 +870,7 @@ private string GetId(object o, bool addToReferencesList)
             string? id = (string?)_references[o];
             if (id == null)
             {
-                id = _idBase + "id" + (++_nextId).ToString(CultureInfo.InvariantCulture);
+                id = $"{_idBase}id{(++_nextId).ToString(CultureInfo.InvariantCulture)}";
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriter.cs:873]:
```suggestion
                id = string.Create(CultureInfo.InvariantCulture, $"{_idBase}id{++_nextId}");
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriter.cs [hunk 10]
```diff
@@ -2835,7 +2834,7 @@ private string GenerateTypeElement(XmlTypeMapping xmlTypeMapping)
 
         private string NextMethodName(string name)
         {
-            return "Write" + (++NextMethodNumber).ToString(null, NumberFormatInfo.InvariantInfo) + "_" + CodeIdentifier.MakeValidInternal(name);
+            return $"Write{(++NextMethodNumber).ToString(null, NumberFormatInfo.InvariantInfo)}_{CodeIdentifier.MakeValidInternal(name)}";
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriter.cs:2837]:
```suggestion
            return string.Create(CultureInfo.InvariantCulture, $"Write{++NextMethodNumber}_{CodeIdentifier.MakeValidInternal(name)}");
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriterILGen.cs [hunk 11]
```diff
@@ -605,7 +605,7 @@ private string GenerateTypeElement(XmlTypeMapping xmlTypeMapping)
 
         private string NextMethodName(string name)
         {
-            return "Write" + (++NextMethodNumber).ToString(null, NumberFormatInfo.InvariantInfo) + "_" + CodeIdentifier.MakeValidInternal(name);
+            return $"Write{(++NextMethodNumber).ToString(null, NumberFormatInfo.InvariantInfo)}_{CodeIdentifier.MakeValidInternal(name)}";
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriterILGen.cs:608]:
```suggestion
            return string.Create(CultureInfo.InvariantCulture, $"Write{++NextMethodNumber}_{CodeIdentifier.MakeValidInternal(name)}");
```

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/QIL/QilValidationVisitor.cs [hunk 12]
```diff
@@ -185,7 +185,7 @@ internal static void SetError(QilNode n, string message)
             string? s = n.Annotation as string;
             if (s != null)
             {
-                message = s + "\n" + message;
+                message = $"{s}\n{message}";
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/QIL/QilValidationVisitor.cs:188]:
```suggestion
                message = $"{s}{Environment.NewLine}{message}";
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationReaderILGen.cs [hunk 13]
```diff
@@ -76,8 +76,8 @@ internal Member(XmlSerializationReaderILGen outerClass, string source, string? a
             internal Member(XmlSerializationReaderILGen outerClass, string source, string? arraySource, string? arrayName, int i, MemberMapping mapping, bool multiRef, string? choiceSource)
             {
                 _source = source;
-                _arrayName = arrayName + "_" + i.ToString(CultureInfo.InvariantCulture);
-                _choiceArrayName = "choice_" + _arrayName;
+                _arrayName = $"{arrayName}_{i.ToString(CultureInfo.InvariantCulture)}";
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationReaderILGen.cs:79]:
```suggestion
                _arrayName = string.Create(CultureInfo.InvariantCulture, $"{arrayName}_{i}");
```

File: src/libraries/System.Private.Xml/src/System/Xml/Xsl/Runtime/XmlQueryOutput.cs [hunk 14]
```diff
@@ -1428,7 +1428,7 @@ private string RemapPrefix(string prefix, string ns, bool isElemPrefix)
             }
 
             // Rule #3: Otherwise, generate a new prefix of the form 'xp_??', where ?? is a stringized counter
-            genPrefix = "xp_" + (_prefixIndex++).ToString(CultureInfo.InvariantCulture);
+            genPrefix = $"xp_{(_prefixIndex++).ToString(CultureInfo.InvariantCulture)}";
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Xsl/Runtime/XmlQueryOutput.cs:1431]:
```suggestion
            genPrefix = string.Create(CultureInfo.InvariantCulture, $"xp_{_prefixIndex++}");
```

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlTextWriter.cs (hunk 3 above)
Comment (kronic) [reply src/libraries/System.Private.Xml/src/System/Xml/Core/XmlTextWriter.cs:1595]:
```cssharp
string.Create(CultureInfo.InvariantCulture, $"d{_top:d}p{temp:d}");
```

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationReader.cs (hunk 7 above)
Comment (kronic) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationReader.cs:799]: span not valid argument;
string? LookupNamespace(string prefix);

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlTextWriter.cs (hunk 3 above)
Comment (kronic) [reply src/libraries/System.Private.Xml/src/System/Xml/Core/XmlTextWriter.cs]: maybe string.Create(CultureInfo.InvariantCulture, $"d{_top}p{temp}";

**Comment by kronic** [comment]
@stephentoub ping

**Comment by eiriktsarpalis** [comment]
@kronic the PR seems to one merge conflict. Would it be possible to rebase your changes on top of the latest main branch.

File: src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationReader.cs (hunk 7 above)
Comment (stephentoub) [reply src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationReader.cs:799]: I thought we had a span-based overload, but I guess not.  You can ignore this suggestion then.

**Comment by eiriktsarpalis** [comment]
Thank you @kronic!

**Pull request merged**